src/board/BoardLogic.java

```diff
@@ -256,6 +257,13 @@ protected void killSnake(int snakeIndex) {
         this.snakesLocation.remove(snakeIndex);
         this.snakes.remove(snakeIndex);
 
+        // check if all existing apples are still valid
+        Set<Field> validAppleFields = getValidAppleFields();
+        for (Field appleField : getApples()) {
+            if (!validAppleFields.contains(appleField)) {
+                removeApple(appleField.getPosX(), appleField.getPosY());
```

Comment:
You may not modify a collection while iterating over it. You could either clone the set, or calculate the fields to be removed as a separate list and remove these elements after the loop has finished